src/Analyzers/CSharp/Analyzers/UseIsNullCheck/CSharpUseNullCheckOverTypeCheckDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,100 @@
+    internal sealed class CSharpUseNullCheckOverTypeCheckDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        private static readonly DiagnosticDescriptor s_useIsNullDescriptor = CreateDescriptorWithId(
+            IDEDiagnosticIds.UseNullCheckOverTypeCheckDiagnosticId,
+            EnforceOnBuildValues.UseNullCheckOverTypeCheck,
+            new LocalizableResourceString(nameof(CSharpAnalyzersResources.Use_is_null_check), AnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)),
+            new LocalizableResourceString(nameof(CSharpAnalyzersResources.Null_check_can_be_clarified), AnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)));
+
+        private static readonly DiagnosticDescriptor s_useIsNotNullDescriptor = CreateDescriptorWithId(
+            IDEDiagnosticIds.UseNullCheckOverTypeCheckDiagnosticId,
+            EnforceOnBuildValues.UseNullCheckOverTypeCheck,
+            new LocalizableResourceString(nameof(CSharpAnalyzersResources.Use_is_not_null_check), AnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)),
+            new LocalizableResourceString(nameof(CSharpAnalyzersResources.Null_check_can_be_clarified), AnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)));
+
+        private static readonly ImmutableArray<DiagnosticDescriptor> s_descriptors = ImmutableArray.Create(s_useIsNullDescriptor, s_useIsNotNullDescriptor);
+
+        public CSharpUseNullCheckOverTypeCheckDiagnosticAnalyzer()
+            : base(s_descriptors)
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(context =>
+            {
+                // All trees should have the same language version. Bail-out early in compilation start instead of checking every tree.
+                var tree = context.Compilation.SyntaxTrees.FirstOrDefault();
+                if (tree is null || ((CSharpParseOptions)tree.Options).LanguageVersion < LanguageVersion.CSharp9)
+                {
+                    return;
+                }
+
+                context.RegisterOperationAction(c => AnalyzeOperation(c), OperationKind.IsType, OperationKind.NegatedPattern);
+            });
+        }
+
+        private static void AnalyzeOperation(OperationAnalysisContext context)
+        {
+            var option = context.Options.GetOption(CSharpCodeStyleOptions.PreferNullCheckOverTypeCheck, context.Operation.Syntax.SyntaxTree, context.CancellationToken);
+            if (!option.Value)
+            {
+                return;
+            }
+
+            if (ShouldReportDiagnostic(context.Operation, out var descriptor) &&
+                context.Operation.Syntax is BinaryExpressionSyntax or UnaryPatternSyntax)
```

Comment:
consider doing the syntax check first.  also, you can register multiple operation actions.  consider if this would simpler to just have two actions registered and handle each case specifically, instead of trying to merge it all into one codepath.